### PR TITLE
dialects: Add LaunchOp and TerminatorOp to GPU.

### PR DESCRIPTION
**Pull request by PapyChacal**
AFAIU, we should be able to run first GPU kernels with those! :slightly_smiling_face: 
I'll start working on an example after that and see if something is missing.

### CODE REVIEW

**Review by webmiche — COMMENTED** [review]
I am confused. Do you actually need to make the operands optional to get the proper operand segment sizes? I feel like this might be an xDSL bug then, that should not be the case, AFAIU.

Or where did you get the information from that they are encoded as being optional?

**Comment by PapyChacal** [comment]
I get how confusing it can be :stuck_out_tongue: 
Thing is, it is an existing MLIR dialect, so if we want interoperability, we want to print it as parsed by MLIR, and parse it as printed by MLIR.
Now, in generic form, MLIR prints and parses gpu.launch as if those arguments are optional, but actually errors if they are not present. It errors too if the perand segment sizes is not there or "wrong" in the optional arguments sense. As it is all auto-generated on our side, I mimic all that this way.
If we want a one-to-one compatibility of the dialect, I think this is the most straightforward way. I guess the decoupled approach would be to have an xDSL GPU dialect and a MLIR GPU dialect we lower to, but that sounds like a second step, if we ever want to take it.

**Comment by PapyChacal** [comment]
Basically, for all those weirdnesses, "where do I get the information it is done that way in MLIR" is trial and error in interfacing with it. I try things first and then implement in xDSL the dialect in the most straightforward MLIR-compatible way I can think of :slightly_smiling_face:

**Comment by webmiche** [comment]
I understand that, sure. What I am saying is that the gpu dialect documentation to me looks like these operands are not optional at all, that's why I am suggesting that you having to declare them as optional to get the operand segment sizes to work might be an xDSL bug in deriving the operand segment sizes.

**Comment by webmiche** [comment]
> Basically, for all those weirdnesses, "where do I get the information it is done that way in MLIR" is trial and error in interfacing with it. I try things first and then implement in xDSL the dialect in the most straightforward MLIR-compatible way I can think of slightly_smiling_face

This sounds even more like it might be an xDSL bug and not  a gpu dialect weirdness, IMO.

**Comment by PapyChacal** [comment]
I don't think we have the same understanding of operand segment sizes then (I'm note sure of mine)
But AFAIU, it contains information about variadic or optional arguments counts only, both in xDSL and MLIR.

So I'm confused, what would be the xDSL bug you're thinking about?

**Comment by PapyChacal** [comment]
But indeed, the GPU dialect doc does not make it sound like those arguments are handled this way.
Though from what I tried with `mlir-opt`, with the commit we use, they are.

**Comment by webmiche** [comment]
Let's take this discussion offline :)
I have opened a [PR](https://github.com/xdslproject/xdsl/pull/491) showing how I think it should work :)

**Comment by PapyChacal** [comment]
For the record or next reader:
My understanding of operand segment sizes was wrong indeed. I thought it was only containing size information for variadics and optionals, but conatins hard 1s for non-variadic non-optionals too.
So there is no hack needed at all! Thanks @webmiche :pray:

**Review by superlopuh — APPROVED** [review]

File: xdsl/dialects/gpu.py [hunk 1]
```diff
@@ -236,6 +249,62 @@ def get() -> LaneIdOp:
...
+
+    def verify_(self) -> None:
+        if len(self.body.blocks) == 0 or all(
+            [len(b.ops) == 0 for b in self.body.blocks]):
+            raise VerifyException("gpu.launch requires a non-empty body.")
+        body_args = self.body.blocks[0].args
+        args_type = [a.typ for a in body_args]
+        if args_type != [IndexType()] * 12:
+            raise VerifyException(
+                f"Expected [12 x {str(IndexType())}], got {[str(t) for t in args_type]}."
```
Comment (webmiche) [inline xdsl/dialects/gpu.py:302]:
```suggestion
                f"Expected [12 x {str(IndexType())}], got {[str(t) for t in args_type]}. "
```

Gives a nicer error message.

Comment (PapyChacal) [reply xdsl/dialects/gpu.py]: Applied, good catch, thanks!

**Review by webmiche — APPROVED** [review]
Nice, LGTM now.

I am very excited to see what you will be able to do with this! :)

**Pull request merged**